### PR TITLE
fix(amplify-provider-awscloudformation): fixed deletion for large bucket

### PR DESCRIPTION
**Pull request by ammarkarachi**
Recursive call to collect and delete large number of objects

#3447, #3451

*Issue #, if available:*

*Description of changes:*
Listing objects and deleting for large buckets

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by kaustavghosh06** [comment]
Could you resolve the conflicts and the fix the circle ci failing test?

File: packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js [hunk 1]
```diff
@@ -67,27 +73,61 @@ class S3 {
...
   }
-
-  deleteAllObjects(bucketName) {
+  getAllObjectKeys(bucketName, continuationToken = null) {
     return new Promise((resolve, reject) => {
       this.s3
-        .listObjects({ Bucket: bucketName })
+        .listObjectsV2({ Bucket: bucketName, ContinuationToken: continuationToken })
         .promise()
         .then((result, lerr) => {
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js]: any reason to not use async / await here? IMO this callback is a little difficult to reason through, especially with the nested promise callback

File: packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js [hunk 2]
```diff
@@ -67,27 +73,61 @@ class S3 {
...
     });
   }
 
+  deleteAllObjects(bucketName) {
+    return new Promise((resolve, reject) => {
+      this.getAllObjectKeys(bucketName).then((result, error) => {
+        if (error) reject(error);
+        const chunkedResult = _.chunk(result, 1000);
+
+        Promise.all(
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js]: nit: more functional to write as:
```javascript
chunkedResult
  .map(res => {
    return {
      Bucket: bucketName,
      Delete: {
        Objects: res,
        Quiet: false,
      }
    }
  )
  .map(delReq => this.s3.deleteObjects(delReq))
  .map(delRes => delRes.promise())
  .reduce((acc, it) => Promise.all(acc, it), Promise.resolve())
  .then(result => resolve(result))
  .catch(err => reject(err));
```

File: packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js [hunk 3]
```diff
@@ -4,18 +4,23 @@ const providerName = require('../../lib/constants').ProviderName;
...
 class S3 {
-  constructor(context, options = {}) {
-    return (async () => {
-      let cred = {};
-      try {
-        cred = await configurationManager.loadConfiguration(context);
-      } catch (e) {
-        // ignore missing config
-      }
+  constructor(context, options = {}, test = false) {
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js]: I don't have a whole lot of context around how this flag is going to be used, but in general "test" flags are a code smell. Could we just have the credentials be an input to the constructor and then it's up to the caller to supply real credentials or test credentials depending on the use case?

**Comment by yuth** [comment]
# [Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3512?src=pr&el=h1) Report
> Merging [#3512](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3512?src=pr&el=desc) into [master](https://codecov.io/gh/aws-amplify/amplify-cli/commit/b6063c520758e69edd8a421e305ccad87a5818e8&el=desc) will **not change** coverage by `%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3512?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##           master    #3512   +/-   ##
=======================================
  Coverage   59.73%   59.73%           
=======================================
  Files         281      281           
  Lines       13136    13136           
  Branches     2642     2642           
=======================================
  Hits         7847     7847           
  Misses       4863     4863           
  Partials      426      426           
```



------

[Continue to review full report at Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3512?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3512?src=pr&el=footer). Last update [b6063c5...b6063c5](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3512?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js (hunk 2 above)
Comment (ammarkarachi) [reply packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.js]: Sounds good I will change that to this

**Pull request merged**